### PR TITLE
Improve the error print of image inspect

### PR DESCRIPTION
**Pull request by miaoyq**
I think we should return the `err` of `GetImage()`, so that user can get the reason why no such image.

Signed-off-by: Yanqiang Miao <miao.yanqiang@zte.com.cn>

### CODE REVIEW

**Comment by duglin** [comment]
This one is kind of interesting.  I couldn't get this error message to appear via the docker cli.  In my testing it appears that this error, while generated, is usually not printed to stderr and instead the cli will generate a new error message in it place.  So this would not benefit the cli user w/o more changes.  Which, btw, might be a good thing to do.

So, looking at the REST API:
```
$curl http://localhost:2375/v1.12/images/qweqweqw/json
No such image: qweqweqw, error: no such id: qweqweqw

$curl http://localhost:2375/v1.12/images/qweqi,weqw/json
No such image: qweqi,weqw, error: Error parsing reference: "qweqi,weqw" is not a valid repository/tag: invalid reference format
```

Here we do see the error with the extra error info - which is good. However, the total error message itself is kind of awkward. The 2nd one in particular makes us sound confused - is there a parsing error or can we not find the image by that name? Since the error is coming from a func called `LookupImage` it might be better to focus on what this func was doing - so, I think it might be better to change the error message to be:
`Error looking for image: ?????, cause: no such id: ????`
`Error looking for image: ?????, cause: Error parsing reference: ...`
Or something like that.

**Comment by thaJeztah** [comment]
We need to be careful; IIRC, `no such` must remain in the response, for our error-code handling

**Comment by duglin** [comment]
First, sigh. :-(
Second, we may want to do more investigation then because if there's a parsing error I wouldn't expect a 404.

**Comment by thaJeztah** [comment]
Yeah, it ain't pretty.

Wrt correct status code, perhaps a 400 or 422

**Comment by miaoyq** [comment]
Thanks @duglin @thaJeztah   I have updated, PTAL.

File: api/server/httputils/errors.go [hunk 1]
```diff
@@ -54,7 +54,7 @@ func GetHTTPErrorStatusCode(err error) int {
 			code    int
 		}{
 			{"not found", http.StatusNotFound},
-			{"no such", http.StatusNotFound},
+			{"no such", http.StatusBadRequest},
```
Comment (thaJeztah) [inline api/server/httputils/errors.go]: This is not correct, because an image/container/network/xyz "not found" is okay to return a 404.

The discussion was if `Error parsing reference: "qweqi,weqw" is not a valid repository....` should return a different code (so only if the specified reference was using an invalid format

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: api/server/httputils/errors.go (hunk 1 above)
Comment (miaoyq) [inline api/server/httputils/errors.go]: Maybe we can add `{"invalid reference format", httphttp.StatusBadRequest},` before `{"no such", http.StatusNotFound}` @thaJeztah

**Comment by duglin** [comment]
While I would prefer if we generated a better error, in order to do that its a much bigger piece of work. For example, this (really bad) pattern of a func returning an error when we can't find something instead of just "nil" for the object we're looking for, makes it really hard to detect when there's a real error (e.g. like the parsing error in this case) vs just a normal (non-error situation) of the object just not being there - which can easily be detected by the 'nil' for the object pointer. Meaning, a non-nil error would be reserved for real errors (non-404's) and we could then act accordingly.  But, this is much bigger than this issue and would require a pretty big change throughout the system.  A good change, IMO, but beyond the scope of this trivial PR.

@miaoyq thanks for the PR and patience.

so, LGTM.

File: daemon/image_inspect.go [hunk 2]
```diff
@@ -14,7 +14,7 @@ import (
 func (daemon *Daemon) LookupImage(name string) (*types.ImageInspect, error) {
 	img, err := daemon.GetImage(name)
 	if err != nil {
-		return nil, fmt.Errorf("No such image: %s", name)
+		return nil, fmt.Errorf("no such image: %s, cause: %v", name, err)
```
Comment (thaJeztah) [inline daemon/image_inspect.go]: should we use `github.com/pkg/errors` instead?

```go
return nil, errors.Wrapf(err, "no such image: %s", name)
```

API error will look like;

```
ERRO[0004] Handler for GET /v1.26/images/foo/json returned error: no such image: foo: no such id: foo
```

Comment (duglin) [inline daemon/image_inspect.go]: no big preference - other than the missing `cause: ` phrase makes a little less hard to parse/read.

Comment (thaJeztah) [inline daemon/image_inspect.go]: True, on the other hand, the original error can then still be obtained through `err.Cause()`

Comment (duglin) [inline daemon/image_inspect.go]: but only on the server, not the end user or api user, right?

Comment (thaJeztah) [inline daemon/image_inspect.go]: yes, correct (unless we modify the code to return it as part of the JSON)

Comment (miaoyq) [inline daemon/image_inspect.go]: Thanks @thaJeztah @duglin   I have updated the code based on the above comments.

**Review by AkihiroSuda — APPROVED** [review]
LGTM if green

**Comment by miaoyq** [comment]
@AkihiroSuda  All checks have passed, PTAL, Thanks

**Comment by thaJeztah** [comment]
@vdemeester could you have a quick peek to check if https://github.com/docker/docker/blob/master/integration-cli/daemon/daemon.go#L764-L767 needs to be updated as well (this PR changes `no such` to be lowercase)

**Review by vdemeester — CHANGES_REQUESTED** [review]
Yep, as @thaJeztah I think we need to update https://github.com/docker/docker/blob/master/integration-cli/daemon/daemon.go#L764-L767 too.

We should use `strings.ToLower` on that check so it works no matter what

**Comment by miaoyq** [comment]
@thaJeztah @vdemeester  Updated!  PTAL  Thanks.

**Review by vdemeester — APPROVED** [review]
LGTM 🐸
